test_runner/regress/test_tenant_detach.py

```diff
@@ -108,6 +108,130 @@ def test_tenant_reattach(
+# This exercises a race condition after tenant attach, where the
+# branch point on the ancestor timeline is greater than the ancestor's
+# last-record LSN. We had a bug where GetPage incorrectly followed the
+# timeline to the ancestor without waiting for the missing WAL to
+# arrive.
+@pytest.mark.parametrize("remote_storage_kind", available_remote_storages())
+def test_tenant_with_branches_reattach_slow_replay(
+    neon_env_builder: NeonEnvBuilder,
+    remote_storage_kind: RemoteStorageKind,
+):
+    neon_env_builder.enable_remote_storage(
+        remote_storage_kind=remote_storage_kind,
+        test_name="test_tenant_reattach",
+    )
+
+    env = neon_env_builder.init_start()
+    env.pageserver.is_testing_enabled_or_skip()
+    pageserver_http = env.pageserver.http_client()
+
+    # Prepare for the test:
+    #
+    # - Main branch, with a table and two inserts to it.
+    # - A logical replication message between the inserts, so that we can conveniently
+    #   pause the WAL ingestion between the two inserts.
+    # - Child branch, created after the inserts
+    # - Detach the tenant
+    tenant_id, timeline_id = env.neon_cli.create_tenant()
+
+    main_endpoint = env.endpoints.create_start("main", tenant_id=tenant_id)
+    with main_endpoint.cursor() as cur:
+        cur.execute("CREATE TABLE test_reattach (t text)")
+        cur.execute("INSERT INTO test_reattach VALUES ('before pause')")
+
+        cur.execute("SELECT pg_logical_emit_message(false, 'neon-test', 'between inserts')")
+
+        cur.execute("INSERT INTO test_reattach VALUES ('after pause')")
+        current_lsn = Lsn(query_scalar(cur, "SELECT pg_current_wal_flush_lsn()"))
+
+    main_endpoint.stop()
+    env.neon_cli.create_branch("child", tenant_id=tenant_id, ancestor_start_lsn=current_lsn)
+
+    pageserver_http.tenant_detach(tenant_id)
+
+    # This fail point will pause the WAL ingestion on the main branch, after the
+    # the first insert
+    before_attach_time = time.time()
+    pageserver_http.configure_failpoints([("wal-ingest-logical-message", "sleep(5000)")])
+
+    # Attach and wait a few seconds to give it time to load the tenants, attach to the
+    # safekeepers, and to stream and ingest the WAL up to the pause-point.
+    pageserver_http.tenant_attach(tenant_id)
+    time.sleep(3)
+
+    # The wal ingestion on the main timeline should now be paused at the fail point.
+    # Run a query on the child branch. The GetPage requests for this should recurse to the
+    # parent timeline, and wait for the WAL to be ingested there. Otherwise it won't see
+    # the second insert.
+    child_endpoint = env.endpoints.create_start("child", tenant_id=tenant_id)
+    with child_endpoint.cursor() as cur:
+        cur.execute("SELECT * FROM test_reattach")
+        assert cur.fetchall() == [("before pause",), ("after pause",)]
+
+    # Sanity check: If this finished in under 5 s, the failpoint didn't sleep correctly.
+    assert time.time() - before_attach_time > 5
```

Comment:
On an overloaded CI runner, this will be true even if we miss the failpoint.

Instead of checking timing, check the pageserver logs for presence / absence of the pg_logical_emit_message that you sent.

`assert not log_contains(the message)` before the attach to ensure we haven't yet processed the message at detach time. (Is this racy because the log is written asynchronously?)

Anyway, more importantly, at this time.time() line, use `wait_until(... log_contains(the message))` instead, to assert that we hit the failpoint.

You can keep the time-based check as a sanity-check, but, the above is more robust.

Reply:
> Instead of checking timing, check the pageserver logs for presence / absence of the pg_logical_emit_message that you sent.

Ok, added.

> `assert not log_contains(the message)` before the attach to ensure we haven't yet processed the message at detach time.

Processing the WAL record doesn't print anything to the log, unless the failpoint is enabled.

> (Is this racy because the log is written asynchronously?)

Yes, it's a little racy. We have that problem with log_contains in general. Would be nice to fix, but not sure how, and hasn't caused any problems in practice so far.

> Anyway, more importantly, at this time.time() line, use `wait_until(... log_contains(the message))` instead, to assert that we hit the failpoint.

The SELECT should not return, until the sleep at the failpoint has finished. So we definitely should have already hit the failpoint at this point, there's no need to wait for it here. Otherwise something is broken.